### ISSUE TITLE
Suspected bug in handling of socks-proxy setting when using UDP

### ISSUE THREAD

**Issue by MixyMisus** [OpenVPN/openvpn#379]
**Describe the bug**

I am running OpenVPN through a SOCKS5 proxy. The connection is established normally and runs okay for minutes to hours.
Periodically the connection tears down with timeout (I am fully confident it is not a network issue) and then OpenVPN attempts restart at which point it hangs for minutes.
After running it with high verbosity logs I found out that while the "hang" part consists of two stages 

stage 1:
Logs do not show anomalies however the software communicating via OpenVPN's tun0 loses connectivity (timeout)

stage 2

For some reason, OpenVPN attempts to reconnect via TCP to the remote OpenVPN server (NOT THE SOCKS PROXY WHICH IS RUNNING ON LOCALHOST)

At this point it can spend minutes (unsuccessfully) trying to reach the remote server directly, finally concludes that network is unreachable and does full restart, which successfully restores connectivity

Here are the [captured logs](https://pastebin.com/YGVDXk9Y) (exclamation marks added by me, interesting part also reproduced in github issue text below)
```

2023-08-02 20:50:17 us=37872 [se-mma-ovpn-018.mullvad.net] Inactivity timeout (--ping-restart), restarting
2023-08-02 20:50:17 us=50012 TCP/UDP: Closing socket
2023-08-02 20:50:17 us=50413 net_route_v4_del: 141.98.255.88/32 via 10.137.5.1 dev [NULL] table 0 metric -1
2023-08-02 20:50:17 us=50774 net_route_v4_del: 0.0.0.0/1 via 10.16.0.1 dev [NULL] table 0 metric -1
2023-08-02 20:50:17 us=50816 net_route_v4_del: 128.0.0.0/1 via 10.16.0.1 dev [NULL] table 0 metric -1
2023-08-02 20:50:17 us=50917 delete_route_ipv6(::/2)
2023-08-02 20:50:17 us=50929 net_route_v6_del: ::/2 via :: dev tun0 table 0 metric -1
2023-08-02 20:50:17 us=51010 delete_route_ipv6(4000::/2)
2023-08-02 20:50:17 us=51020 net_route_v6_del: 4000::/2 via :: dev tun0 table 0 metric -1
2023-08-02 20:50:17 us=51056 delete_route_ipv6(8000::/2)
2023-08-02 20:50:17 us=51067 net_route_v6_del: 8000::/2 via :: dev tun0 table 0 metric -1
2023-08-02 20:50:17 us=51103 delete_route_ipv6(c000::/2)
2023-08-02 20:50:17 us=51111 net_route_v6_del: c000::/2 via :: dev tun0 table 0 metric -1
2023-08-02 20:50:17 us=51149 Closing TUN/TAP interface
2023-08-02 20:50:17 us=51166 net_addr_v4_del: 10.16.0.3 dev tun0
2023-08-02 20:50:17 us=51544 net_addr_v6_del: fdda:d0d0:cafe:1302::1001/64 dev tun0
2023-08-02 20:50:17 us=57855 SIGUSR1[soft,ping-restart] received, process restarting
2023-08-02 20:50:17 us=59705 Restart pause, 1 second(s)
2023-08-02 20:50:18 us=62325 Re-using SSL/TLS context
2023-08-02 20:50:18 us=65480 Control Channel MTU parms [ mss_fix:0 max_frag:0 tun_mtu:1250 tun_max_mtu:0 headroom:126 payload:1600 tailroom:126 ET:0 ]
2023-08-02 20:50:18 us=65513 Data Channel MTU parms [ mss_fix:0 max_frag:0 tun_mtu:1450 tun_max_mtu:1600 headroom:136 payload:1800 tailroom:568 ET:32 ]
2023-08-02 20:50:18 us=65814 TCP/UDP: Preserving recently used remote address: [AF_INET]141.98.255.88:1302
2023-08-02 20:50:18 us=65856 Socket Buffers: R=[212992->425984] S=[212992->425984]
!!!!!!!!!!!!!!!
2023-08-02 20:55:11 us=65866 Attempting to establish TCP connection with [AF_INET]141.98.255.88:1302
!!!!!!!!!!!!
2023-08-02 20:55:11 us=66077 TCP: connect to [AF_INET]141.98.255.88:1302 failed: Network is unreachable
2023-08-02 20:55:11 us=66138 TCP/UDP: Closing socket
2023-08-02 20:55:11 us=66173 SIGUSR1[connection failed(soft),connection-failed] received, process restarting
2023-08-02 20:55:11 us=66259 Restart pause, 1 second(s)
2023-08-02 20:55:12 us=66354 Re-using SSL/TLS context
2023-08-02 20:55:12 us=66448 Control Channel MTU parms [ mss_fix:0 max_frag:0 tun_mtu:1250 tun_max_mtu:0 headroom:126 payload:1600 tailroom:126 ET:0 ]
2023-08-02 20:55:12 us=66466 Data Channel MTU parms [ mss_fix:0 max_frag:0 tun_mtu:1450 tun_max_mtu:1600 headroom:136 payload:1800 tailroom:568 ET:32 ]
2023-08-02 20:55:12 us=66499 TCP/UDP: Preserving recently used remote address: [AF_INET]127.0.0.1:2069
2023-08-02 20:55:12 us=66541 Socket Buffers: R=[212992->425984] S=[212992->425984]
2023-08-02 20:55:12 us=66550 Attempting to establish TCP connection with [AF_INET]127.0.0.1:2069
2023-08-02 20:55:12 us=66618 TCP connection established with [AF_INET]127.0.0.1:2069
2023-08-02 20:55:12 us=67384 SOCKS proxy wants us to send UDP to [AF_INET]127.0.0.1:2069
2023-08-02 20:55:12 us=67401 NOTE: setsockopt TCP_NODELAY=1 failed
2023-08-02 20:55:12 us=67414 UDPv4 link local: (not bound)
2023-08-02 20:55:12 us=67423 UDPv4 link remote: [AF_INET]185.213.154.138:1302
2023-08-02 20:55:13 us=58053 TLS: Initial packet from [AF_INET]185.213.154.138:1302, sid=56293a44 16504030
2023-08-02 20:55:13 us=105546 VERIFY OK: depth=2, C=SE, ST=Gotaland, L=Gothenburg, O=Amagicom AB, OU=Mullvad, CN=Mullvad Root CA v2, emailAddress=security@mullvad.net
```

**To Reproduce**
Start OpenVPN in UDP mode with SOCKS5 enabled (socks-proxy in config, so socks proxy must support UDP. Ideally it should be on localhost since in my setup it is)

Run it for a while, with moderate load.

P.S.
Upon further experiments I discovered that the "strange attempt to bypass socks proxy" also happens upon "normal" connectivity loss events (power cycling the wifi router to which the machine with VPN client is connected)

**Expected behavior**
Everything to work without weird connectivity losses.

If a connectivity loss happens for normal reasons do not perform useless attempt to connect to remote server bypassing the socks proxy and re-connect to socks proxy (which it eventually does but only after very long wait)

**Version information (please complete the following information):**

 - OS: Debian 11 (bare metal and VM), Debian 10 (VM), Fedora (VM) (issue reproduced on all)
 
Tested on two following OpenVPN versions (reproduces):

OpenVPN 2.4.7 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Apr 28 2021
library versions: OpenSSL 1.1.1n  15 Mar 2022, LZO 2.10
 
OpenVPN 2.6.5 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] [DCO]
library versions: OpenSSL 3.1.1 30 May 2023, LZO 2.10
 

**Additional context**
Socks capability is why I'm using openvpn in the first place so any suggestion on how to work around this would be extremely welcome

**Comment by cron2** [comment]
Your description definitely sounds like a bug (and a long-standing one at that).

I tried to reproduce this with current master, connecting to a (non-local) SOCKs proxy, and then onward to a TCP server.  Triggering the two different kind of reconnects with `SIGUSR1` or `SIGHUP` did not do anything unusual (so it's not totally easily reproduced).  Will try timeouting and network errors next.

**Comment by cron2** [comment]
"Killing the SOCKS proxy" is also not sufficient, this will produce a correct reconnect towards the SOCKS proxy (and as soon as it's back, proper re-establishment).  So, network errors between proxy and server, and timeout next...

**Comment by cron2** [comment]
Mmmh, this is going to be a tougher nut.  Tried `--keepalive 10 60` on the client, without `--ping` on the server, which triggers a timeout restart after 60 seconds without user traffic

```
2023-08-03 12:26:49 SIGUSR1[soft,ping-restart] received, process restarting
2023-08-03 12:26:49 Restart pause, 1 second(s)
2023-08-03 12:26:50 TCP/UDP: Preserving recently used remote address: [AF_INET]194.97.144.213:1080
2023-08-03 12:26:50 Socket Buffers: R=[65536->65536] S=[32768->32768]
2023-08-03 12:26:50 Attempting to establish TCP connection with [AF_INET]194.97.144.213:1080
2023-08-03 12:26:50 TCP connection established with [AF_INET]194.97.144.213:1080
2023-08-03 12:26:50 TCPv4_CLIENT link local: (not bound)
2023-08-03 12:26:50 TCPv4_CLIENT link remote: [AF_INET]194.97.144.213:1080
2023-08-03 12:26:50 TLS: Initial packet from [AF_INET]194.97.144.213:1080, sid=1af58c31 a8b415b8
```

... which nicely reconnects to the SOCKs proxy...

@MixyMisus I wonder if we're doing something special to "localhost" here.  Can you reproduce if the SOCKS proxy is accessed via the LAN IP (so, still running on the same machine, but no "127.0.0.1" in internal data structures)?  Wild guess, though...

**Comment by cron2** [comment]
ooooh wait.  I was misreading the UDP part, and using a TCP server.  Let me retry.

**Comment by MixyMisus** [comment]
@cron2 Ha! Caught a real bug! Lucky me! 

BTW there appears to be one more bug / misbehavior when using OpenVPN(UDP)+SOCKS5 proxy but I only manage to get it like once a day (both with LAN SOCKS5 server and localhost SOCKS5) and never got a conclusive log for it (it's bad, I know :))

Should I create a separate issue for it ?

**Comment by LindaFerum** [comment]
Hi! Just found this thing via googling around and reporting that 1) it reproduces on all my OpenVPN installs and 2) it is somewhat disruptive to my stuff. I understand there are higher priority concerns and this is open source product so like, no hard feelings on my part, just leaving my two cents here hoping this bug will be picked up and fixed one day 

Thanks :)

**Comment by 5andr0** [comment]
## TL;DR

1. Reconnecting to VPN remote as proxy instead of socks ip after `SIGUSR1[soft,ping-restart]`


   | Cause                                                                                                                                                                                                                                                                                                                                                                                                | Fix                                                                                                                                                                                                                                                           |
   | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
   | Unlike other proxy protocols with proxy-based DNS resolution, SOCKS5 UDP ASSOCIATE uses local DNS for resolving the VPN server domain, causing the cached IP to be mixed with the SOCKS server IP in the remote address list, which persists across soft restarts ([init.c#L3991](https://github.com/OpenVPN/openvpn/blob/2cc00e80bca26c1c7129fa45e48714ca68b56b28/src/openvpn/init.c#L3991https:/)) | <ul><li>Directly connect to proxy servers without cached remote addr</li><li>Always clear `remote_addrlist` for udp proxy</li><li>Make use of socks5 *ATYP=03* (Domain) for udp on each packet (Lots of overhead, but preserves privacy over proxy)</li></ul> |
2. Timeouts and Ping restarts


   | Cause                                                                                                                                                                                                                                                                                                                | Fix                                            |
   | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------ |
   | A UDP association terminates when the TCP connection that the UDP ASSOCIATE request arrived on terminates ([RFC 1928](https://datatracker.ietf.org/doc/html/rfc1928#section-6:~:text=A%20UDP%20association%20terminates))<br />There's no TCP Keepalive, which causes NAT timeouts, firewall drops, idle disconnects | Add TCP Keepalive for`sock->ctrl_sd`           |
   | Tunneling through socks or chained VPNs adds overhead, lowering the effective MTU and causing fragmentation. Fragmented datagrams are dropped silently, not resetting the ping timer and causing connection instability                                                                                              | Lower values for ***tun-mtu*** and ***mssfix*** |

# Detailed Analysis

Thanks @MixyMisus for reporting this. It's unfortunate that this issue has been stale for years, given that UDP is the only sensible option for tunneling OpenVPN.

#### Disclaimer

I'm not an OpenVPN dev! While I've written an ovpn patch before, I don't have time to implement or test this atm. There's a discussion needed anyway on how to fix this. I'll do my best to provide detailed information so you guys can fix it.

I only spent an afternoon digging through the code to get to the bottom of this, without any testing. This might just be the tip of the iceberg, with more bugs still lurking.

Would be nice to have your help @cron2 to test this.

## Bug #1 Reconnecting to VPN remote as proxy instead of socks ip after `SIGUSR1[soft,ping-restart]`

Thanks to the logs, the root cause is fairly clear. During a soft restart, the remote address list is not cleared, causing the previously resolved VPN IP to be reused when reconnecting to the SOCKS5 server.

This issue only affects SOCKS5 UDP ASSOCIATE connections, since all other proxy methods resolve VPN remotes via proxy DNS.

⚠️ As a result, `persist-remote-ip` is effectively ignored. A clear warning should be implemented whenever a proxy is used in combination with `persist-remote-ip`.

For SOCKS5 UDP ASSOCIATE, the VPN remote is resolved locally and incorrectly merged into the remote address list together with the SOCKS5 server IP, which causes this bug on soft restarts.

⚠️ This MUST trigger a prominent warning when UDP ASSOCIATE is used. Users reasonably expect their privacy preserved when routing through a proxy. This behavior creates an unintended privacy leak by exposing the VPN remote outside the SOCKS5 path. Although SOCKS5 itself is unencrypted, most users will have socks5 tunneled for this exact reason.

There's a few options to tackle this:

1. Always clear remote addrlist for udp socks (disables persist-remote-ip)

   [init.c#L3991](https://github.com/OpenVPN/openvpn/blob/2cc00e80bca26c1c7129fa45e48714ca68b56b28/src/openvpn/init.c#L3991https:/)

```c
if (!(c->sig->signal_received == SIGUSR1
      && !(c->c1.socks_proxy != NULL 
          && c->options.ce.proto == PROTO_UDP)  
          /* Always clear remote_list for SOCKS UDP */
      && ((c->options.persist_remote_ip)
          || (c->sig->source != SIG_SOURCE_HARD
              && ((c->c1.link_socket_addrs[0].current_remote
                   && c->c1.link_socket_addrs[0].current_remote->ai_next)
                  || c->options.no_advance)))))
{
    clear_remote_addrlist(&c->c1.link_socket_addrs[0], !c->options.resolve_in_advance);
}
```

2. Directly connect to all proxy servers, not using vpn persist-remote-ip cache.
   Can be implemented with some booleans and *if* in `resolve_remote`
3. The privacy-preserving approach would be to use **ATYP=03 (domain)** instead of IPv4 in the SOCKS5 protocol.

   The drawback is that the VPN hostname must be included in every UDP packet header, which introduces significant overhead and may require dynamic MTU adjustments. However, for privacy reasons, this should still be offered as an optional feature.

   To mitigate this overhead while preserving proxy privacy, the hostname can be resolved in advance via the SOCKS5 UDP proxy by performing a raw DNS query to a configured DNS server.

   **Bonus:** When implemented correctly, `persist-remote-ip` would work with this method.

## Bug #2 Timeouts and Ping restarts

Without logs, it's difficult to pinpoint the exact cause. However, I see two likely suspects:

1. According to the specification [RFC 1928](https://datatracker.ietf.org/doc/html/rfc1928#section-6:~:text=A%20UDP%20association%20terminates), the UDP relay terminates when the TCP control connection is closed. Once the UDP relay is established, the control connection typically remains idle. Without TCP keepalives, this connection may be terminated due to NAT timeouts, firewall policies, or operating system settings.

   The solution is to add `TCP_KEEPALIVE` values to `sock->ctrl_sd` and monitor `ctrl_sd` in the event loop `event_ctl(es, ctrl_sd, EVENT_READ, ...);`

   On an unexpected close event I would add a log output and a SIGUSR1 signal.

   ```c
   msg(M_WARN, "SOCKS TCP control connection closed, restarting");
   register_signal(c->sig, SIGUSR1, "socks-tcp-closed");
   ```
2. Tunneling through socks or chained VPNs adds overhead, lowering the effective MTU and causing fragmentation. Fragmented datagrams are dropped silently, failing to reset ping timers and resulting in connection instability.

   While this is unlikely to cause ping restarts —since ping packets themselves should not be dropped—it is still worth mentioning. Without adjusting MTU values to account for tunnel overhead, this will almost certainly lead to stability issues.

   For testing just go for

   ```
   tun-mtu 1300
   mssfix 1250
   ```

## Workarounds until fixed

Exclude **persist-remote-ip** from the config or use ips instead of hostnames.

A local proxy tcp connection should be kept alive when your os defaults are correct.

Socat and HAProxy can forward local TCP with adding keepalives towards remote socks server. Something like:
`socat TCP-LISTEN:2069,fork,reuseaddr,keepalive,keepidle=60,keepintvl=30,keepcnt=3 TCP:real-socks-server:1080`
